### ISSUE TITLE
Consider Tor Multi-vm/machine Designs

### ISSUE THREAD

**Issue by HulaHoopWhonix** [Jesse-V/OnioNS-server#53]
Jessie please keep designs like Whonix in mind where Tor is not running on the same machine as TBB. There is a Gateway VM where Tor runs and where we plan to add OnioNS standalone package so it can process requests. What we'll need is a boolean setting for the OnioNS thats bundled with TBB so it can be disabled and only the instance on the gateway is used.

**Comment by Jesse-V** [comment]
Hi, realize that I have relatively recently split the software into four and made different packages. Currently OnioNS-HS package has a dependency on system-wide Tor, OnioNS-server has no dependency on any Tor, and OnioNS-client needs the Tor Browser. As I understand your request, would this design resolve this situation?

**Comment by HulaHoopWhonix** [comment]
Sorry for not replying sooner, things got in the way.

If I understand correctly, the architecture components and their functions:

*OnioNS-Server - run by Quorum members
*OnonNS-Client - the TBB integrated client that queries Quorum
*OnioNS-HS - run on the same machine as the Tor instance of an Onion Service

Since we cover both client and HS usecases I'll discuss both.

OnioNS-HS should work with Whonix by running on the Gateway VM.

OnonNS-Client would need to drop its dependency on TBB and instead work with system-wide Tor on the Gateway VM so it can process DNS requests on behalf of potentially many different types of applications that could run on the Workstation VM behind it (and not just TBB).

**Comment by Jesse-V** [comment]
Indeed. I do have plans on supporting things other than the Tor Browser, but at the moment that's not happening. OnioNS-client doesn't have a hard dependency on the Tor Browser, but I did build it with the integration in mind. However, if you prefer, you may modify the code in main.cpp or Client.cpp to bind to SOCKS port 9050 instead of 9150, hence switching to system-wide Tor.

**Comment by HulaHoopWhonix** [comment]
>Indeed. I do have plans on supporting things other than the Tor Browser, but at the moment that's not happening. OnioNS-client doesn't have a hard dependency on the Tor Browser, but I did build it with the integration in mind. However, if you prefer, you may modify the code in main.cpp or Client.cpp to bind to SOCKS port 9050 instead of 9150, hence switching to system-wide Tor.

OK, When you get around to implementing this can you please do it so its configurable via a settings file. We only ship binaries that identical to upstream's for security and support reasons and so patching the code directly can't be an option.

**Comment by Jesse-V** [comment]
That's understandable. I'll rename this ticket to indicate that I should provide a flag for using a specific SOCKS port, that way it can connect to different Tor instances as needed.

**Comment by adrelanos** [comment]
Here is an overview how Tor Browser works in Whonix:

* Whonix does not modify Tor Browser's startup script, defaults, etc. (This is useful to be able to provide easy manual [instructions](https://www.whonix.org/wiki/Manually_Updating_Tor_Browser) for updating Tor Browser. Ones, that do not result in Tor over Tor. In past download locations and whatnot were often changed by The Tor Project, so that any update scripts would break in meanwhile.)
* In Whonix-Workstation.
 * rinetd listens on 127.0.0.1 9150 and 9151 (TBB's default ports) and forwards them to Whonix-Gateway 10.152.152.10 9150 (where a Tor SocksPort is listening) and 9151 (where [Control Port Filter Proxy](https://www.whonix.org/wiki/Dev/CPFP) is listening).
 * Tor does not get started by the tor-launcher Firefox add-on because the [TOR_SKIP_LAUNCH](https://trac.torproject.org/projects/tor/ticket/6009) environment variable has been set set to 1.

A dependency on the Debian `tor` package would not be that bad, due to [anon-ws-disable-stacked-tor](https://www.whonix.org/wiki/Dev/anon-ws-disable-stacked-tor).

There are no testers builds of TBB yet that are using OnioNS? Without knowing who is listening where and how OnonNS-Client will be distributed (Debian repository etc.) it's hard to make suggestions on how it could be made work with Whonix.

I guess OnonNS-Client will be a DNS resolving server service opening a local listening port? If you could describe please how TBB integration will work, who/who services are started, listening where, connecting where, then maybe we can find an equally awesome solution.

**Comment by Jesse-V** [comment]
I'd be happy to explain. OnioNS-client opens 9053/TCP on localhost which listens for *.tor domains, and then writes *.onion back. A Python script, using Stem, receives event notifications through Tor's control port of stream requests, picks up a *.tor domain lookup, sends it to 9053/TCP for resolution, reads back a *.onion address, tells Tor over the control port to redirect that stream to that *.onion address, and then tells Tor to attach that stream to a circuit. In effect, the Tor Browser operates exactly as before and from its perspective the user is browsing a *.tor domain, when under the hood Tor is receiving a *.onion lookup. This has the effect of loading hidden services under *.tor domain names.

OnioNS is mainly distributed through my Ubuntu PPA, as described in the README. My current approach for integrating with the Tor Browser is to rename the Tor executable in Browser/TorBrowser/Data to "torbin", then insert my own executable which launches torbin, /usr/bin/onions-client, and the Python Stem script as child processes, in that order. That way, all the necessary software starts when the Tor Browser opens and by monitoring the status of torbin, they can all close when the Tor Browser closes as well. Obviously, this setup is not ideal and I do have plans to modify the Tor Launcher to also start up these processes, but my binary substitution approach was quicker and easier.

In summary, I'm not distributing a different build of the Tor Browser. Clients can still download and verify it via PGP and all that, but then they can get OnioNS functionality into it with a few rename and copy commands. That's my current approach anyway. I'd love to integrate with Whonix of course, but I would like to prepare working binaries of non-Whonix software first, and then we can explore integrating with Whonix; one battle at a time. It's entirely possible that it'll be easy to integrate in Whonix, in which case we've killed two birds with one stone.
